cpp/src_prims/distance/distance.h

```diff
@@ -326,7 +326,7 @@ void pairwiseDistance(const Type *x, const Type *y, Type *dist, Index_ m,
  *                  the epsilon neighborhood.
  */
 template <DistanceType distanceType, typename T, typename Lambda,
-          typename Index_ = int, typename OutputTile_ = OutputTile_8x128x128>
+          typename Index_ = long, typename OutputTile_ = OutputTile_8x128x128>
```

Comment:
Do we need to change this default value? This can cause other algos relying on this prim to loose perf due to this.

Reply:
Yep. I’m really happy we can now change the integral precision just by changing the index type. I realized the main problem for int was that the N*batchSize was overflowing when N grew into the 100s of millions. 

For now, I’ve changed the python back to using int32, and I’ve added an additional conditional to the computation of the batch size that will make sure the batch size won’t overflow. It turns out, in some cases, for smaller values of N at least, doing more smaller batches with int32 can still be faster than fewer larger batches with long.

I think we should expose an argument in the DBSCAN Python that will allow the user to set the precision of the output. Perhaps we could even provide an “auto” setting that will automatically set it to single or double precision based on the tipping point (when it becomes reasonable to use double precision index type). I’ll do that in this PR and maybe set the tipping point at the 10s of millions for now. At 1M, it still seems that int can be faster, even with many more batches and less average GPU utilization.

Reply:
Added `out_dtype` argument to `fit()` and associated pytests. Also tagging @dantegd